### ISSUE TITLE
How to build a yolo pipeline?

### ISSUE THREAD

**Issue by ritwikdubey** [AlexeyAB/darknet#1492]
Hi @AlexeyAB ,
Is it possible to build a detection pipeline within darknet? 
Say in stage 1 I want to detect all faces in an image and then in stage 2 I want to perform face recognition of a known or unknown person.

Right now I perform face recognition on the whole image but I wonder if I use multistage detection will it improve the accuracy?

**Comment by AlexeyAB** [comment]
@ritwikdubey Hi,

There is no multi-stage prediction. 
* Or implement multi-stage prediction by yourself inside Darknet-code
* Or implement it in your own C/C++-code by using darknet.so/dll library

Yes, multi-stage inference can improve accuracy, if you will crop faces from original not-resized image and recognize each face in separate inference by using neural network that is optimal for face recognition.

**Comment by Muhammad057** [comment]
Hello @ritwikdubey, I am also looking into the same problem these days. I have two separate YOLO models. One for License Plate Detection from an incoming vehicle, and secondly, the other model for recognizing the license plate from the detected model. I need to build a pipeline, where the output of detected model is fed to the recognized model. If you have any further insight, kindly share with me. Thank you.

**Comment by Muhammad057** [comment]
Hello @AlexeyAB, regarding your comment 'implement multi-stage prediction by yourself inside Darknet-code', can you please point me out to that file, where I have to make changes for multi-stage prediction? Thank you.

**Comment by AlexeyAB** [comment]
@Muhammad057 
To implement it for Detection on the images you should modify this function: https://github.com/AlexeyAB/darknet/blob/18d5e4f39c1441f2c21043ac3204b5cb279f8758/src/detector.c#L1083

----

1. For example, add here: https://github.com/AlexeyAB/darknet/blob/18d5e4f39c1441f2c21043ac3204b5cb279f8758/src/detector.c#L1090
this code - where is resnet152.cfg/weights - your model for classification (face recognition, ...):
```cpp
    network net_classifier = parse_network_cfg_custom("cfg/resnet152.cfg", 1);
    load_weights(&net, "resnet152.weights");
```

----

2. And add between these lines: https://github.com/AlexeyAB/darknet/blob/18d5e4f39c1441f2c21043ac3204b5cb279f8758/src/detector.c#L1142-L1143
the code for second stage recognition - something like this (may be I forgot something):
```cpp
            for (i = 0; i < nboxes; ++i) {
                int class_id = -1;
                float prob = 0;
                for (j = 0; j < l.classes; ++j) {
                    if (dets[i].prob[j] > thresh && dets[i].prob[j] > prob) {
                        prob = dets[i].prob[j];
                        class_id = j;
                    }
                }
                if (class_id >= 0) {
                    image im_classify = crop_image(im, dets[i].bbox.x, dets[i].bbox.y, dets[i].bbox.w, dets[i].bbox.h);
                    image r = letterbox_image(im_classify, net_classifier.w, net_classifier.h);
                    float *predictions_classify = network_predict(net_classifier, r.data);

                    int top = 1;
                    top_k(predictions_classify , net_classifier.outputs, top, indexes);
                    int *indexes = calloc(top, sizeof(int));

                    for(i = 0; i < top; ++i){
                        int index = indexes[i];
                         printf("%s: %f\n",names[index], predictions_classify [index]);
                    }
                }
            }
```

**Comment by ritwikdubey** [comment]
I believe you meant `net_classifier` in `load_weights` function.

``
   network net_classifier = parse_network_cfg_custom("cfg/resnet152.cfg", 1);
    load_weights(&net, "resnet152.weights");
``

**Comment by AlexeyAB** [comment]
> I believe you meant net_classifier in load_weights function.

Yes, I fixed it.

> BTW, why letterbox_image the image? I can't open the code files right now but I guess this function draws the colored box around the prediction and adds label to it, correct? Why would I want to add that before sending the cropped image for second prediction?

No, it doesn't draw boxes. `letterbox_image()` resizes image. More about `letterbox_image()` and `resize_image()`: https://github.com/AlexeyAB/darknet/issues/232#issuecomment-336955485

**Comment by Muhammad057** [comment]
@AlexeyAB 
Thanks for the reply.  Ok, I have added the above code and recompiled the darknet.sln file. When I run the darknet.exe file, it takes the .cfg file and weights file from your below mentioned code to detect the license plate from a vehicle.

net_classifier = parse_network_cfg_custom("cfg/yolo-obj.cfg", 1);          //yolo_obj is the test cfg file
load_weights(&net_classifier , "yolo-obj_4600.weights");                       //trained weights file

But, what I wanted to do is a little bit different. I run the trained YOLO on a live stream (rtsp link) through this command 'darknet.exe detector demo data/obj.data cfg/yolo-obj.cfg backup/yolo-obj_4600.weights rtsp link', and I save the frames in 'result_img' folder. Now, I want to test each frame through another trained model (recognition phase). Can you please help me to point out the relevant changes that I need to do in src/detector.c file? Where can I add the recognition model's cfg file, weights file, obj.data & obj.names file? 
Regards.

**Comment by ritwikdubey** [comment]
@Muhammad057 
Good to know. 

> One for License Plate Detection from an incoming vehicle, and secondly, the other model for recognizing the license plate from the detected model. I need to build a pipeline, where the output of detected model is fed to the recognized model. If you have any further insight, kindly share with me. Thank you.
I sorta worked on similar problem. I found out that Yolo v3 was quite good in detecting numbers off license plate without character segmentation stage. I guess the way you're thinking is sufficient; 1- localize and segment the license plate region 2- Run prediction on the scaled license plate to detect the numbers. 
![predictions](https://user-images.githubusercontent.com/1186630/44930817-23c04b00-ad25-11e8-9e34-4558af82e1b6.jpg)

**Comment by Muhammad057** [comment]
@ritwikdubey 
Yes, I have already detected the numbers from the license plate (which I call it my recognition model) using YOLOV3. But before detecting the license plate (LP) numbers, first, I segment out the license plate from a vehicle through my detection model and then, fed it to the recognition model (as started in my earlier comments). I think, YOLO doesn't have any pipeline for multi-stage prediction , i.e, first detect the LP and then recognize the letters from the same src/detector.c code. Below is the snippet of detecting the numbers from a LP 
![recognized_image431 - copy](https://user-images.githubusercontent.com/40855134/44942877-def3ed00-add4-11e8-8c57-76c188f4f12e.jpg)

**Comment by kmsravindra** [comment]
> ```c++
>     network net_classifier = parse_network_cfg_custom("cfg/resnet152.cfg", 1);
>     load_weights(&net_classifier , "resnet152.weights");
> ```

@AlexeyAB , Hi

If I have a custom classifier (like inceptionV2) implemented in keras that takes the bounding box crops as image inputs. So, is it possible to plug that classifier on top of darknet detection? If so, could you please tell me where do I build that model , compile it and load the trained weights for the same? Thanks for your help!

**Comment by Favi0** [comment]
> @Muhammad057
> Good to know.
> 
> > One for License Plate Detection from an incoming vehicle, and secondly, the other model for recognizing the license plate from the detected model. I need to build a pipeline, where the output of detected model is fed to the recognized model. If you have any further insight, kindly share with me. Thank you.
> 
> I sorta worked on similar problem. I found out that Yolo v3 was quite good in detecting numbers off license plate without character segmentation stage. I guess the way you're thinking is sufficient; 1- localize and segment the license plate region 2- Run prediction on the scaled license plate to detect the numbers.
> ![predictions](https://user-images.githubusercontent.com/1186630/44930817-23c04b00-ad25-11e8-9e34-4558af82e1b6.jpg)

what do you mean by scaled?

**Comment by barzan-hayati** [comment]
@Muhammad057 Are you able to use pipeline in order to combine two stage of plate detection and recognition? I found an article that did license plate recognition by one model not two.

[Real-Time Brazilian License Plate Detection and Recognition Using Deep Convolutional Neural Networks](http://www.inf.ufrgs.br/~smsilva/real-time-brazilian-alpr/)

**Comment by Muhammad057** [comment]
@barzan-hayati Yes I created pipeline to combine both of my models. Anyways thanks for sharing.

**Comment by barzan-hayati** [comment]
@Muhammad057 Could you please explain your method for pipeline?
